File: 152-maximum-product-subarray/152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        if (nums.size() == 0)
            return 0;
        
        int Max = nums[0], Min = nums[0], result = nums[0];
        
        for (int i = 1; i < nums.size(); i++) {
            int temp = Max;
            Max = max(max(Max * nums[i], Min * nums[i]), nums[i]);
            Min = min(min(temp * nums[i], Min * nums[i]), nums[i]);
            
            if (result < Max)
                result = Max;
        }
        return result;
    }
};

File: 152-maximum-product-subarray/152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        if (nums == null || nums.length == 0)
            return 0;
        
        int max = nums[0], min = nums[0], result = nums[0];
        for (int i = 1; i < nums.length; i++) {
            int temp = max;
            max = Math.max(Math.max(max * nums[i], min * nums[i]), nums[i]);
            min = Math.min(Math.min(temp * nums[i], min * nums[i]), nums[i]);
            
            if (max > result)
                result = max;
        }
        return result;
    }
}

// bruteforce
// class Solution {
//     public int maxProduct(int[] nums) {
//         int max = Integer.MIN_VALUE, pro = nums[0];
//         for (int i = 0; i < nums.length; i++) {
//             pro = 1;
//             for (int j = i; j < nums.length; j++) {
//                 pro *= nums[j];
//                 max = Math.max(max, pro);
//             }
//         }
//         return max;
//     }
// }